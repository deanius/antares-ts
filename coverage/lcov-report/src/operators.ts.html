<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/operators.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../index.html">All files</a> / <a href="index.html">src</a> operators.ts
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">26.92% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>7/26</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/12</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/10</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">33.33% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>7/21</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a>
<a name='L44'></a><a href='#L44'>44</a>
<a name='L45'></a><a href='#L45'>45</a>
<a name='L46'></a><a href='#L46'>46</a>
<a name='L47'></a><a href='#L47'>47</a>
<a name='L48'></a><a href='#L48'>48</a>
<a name='L49'></a><a href='#L49'>49</a>
<a name='L50'></a><a href='#L50'>50</a>
<a name='L51'></a><a href='#L51'>51</a>
<a name='L52'></a><a href='#L52'>52</a>
<a name='L53'></a><a href='#L53'>53</a>
<a name='L54'></a><a href='#L54'>54</a>
<a name='L55'></a><a href='#L55'>55</a>
<a name='L56'></a><a href='#L56'>56</a>
<a name='L57'></a><a href='#L57'>57</a>
<a name='L58'></a><a href='#L58'>58</a>
<a name='L59'></a><a href='#L59'>59</a>
<a name='L60'></a><a href='#L60'>60</a>
<a name='L61'></a><a href='#L61'>61</a>
<a name='L62'></a><a href='#L62'>62</a>
<a name='L63'></a><a href='#L63'>63</a>
<a name='L64'></a><a href='#L64'>64</a>
<a name='L65'></a><a href='#L65'>65</a></td><td class="line-coverage quiet"><span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import { Observable, from, of } from "rxjs"
import { ajax } from "rxjs/ajax"
import { StreamingGetOptions } from "./types"
import { delay, map, flatMap } from "rxjs/operators"
import { compare, Operation } from "fast-json-patch"
&nbsp;
/** @description Delays the occurrence of an object, or the
 * invocation of a function, for the number of milliseconds given
 * @returns An Observable of the desired effect/object
 * @example after(100, {type: 'Timedout'}).subscribe(action =&gt; ...)
 */
export const after = <span class="fstat-no" title="function not covered" >(</span>ms: number, objOrFn: Object | Function): Observable&lt;any&gt; =&gt; {
  const <span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="fstat-no" title="function not covered" >[<span class="cstat-no" title="statement not covered" >obj,</span> <span class="cstat-no" title="statement not covered" >effect]</span> =</span></span></span>
    objOrFn instanceof Function ? [null, objOrFn] : [objOrFn, (value: Object) =&gt; value]
&nbsp;
<span class="cstat-no" title="statement not covered" >  return of(obj).pipe(</span>
    delay(ms),
    // @ts-ignore
    map(effect)
  )
}
&nbsp;
/** @description Gets the resource, returning an Observable of resources referred to by the URL
 * @see https://medium.com/@deaniusaur/how-to-stream-json-data-over-rest-with-observables-80e0571821d3
 */
export const ajaxStreamingGet = <span class="fstat-no" title="function not covered" >(</span>opts: StreamingGetOptions): Observable&lt;any&gt; =&gt; {
  // An Observable of the response, expanded to individual results if applicable.
<span class="cstat-no" title="statement not covered" >  return ajax({</span>
    url: opts.url,
    method: opts.method || "GET",
    withCredentials: Boolean(opts.withCredentials),
    timeout: opts.timeout || 30 * 1000
  }).pipe(
    // @ts-ignore
    flatMap(<span class="fstat-no" title="function not covered" >ajax =&gt; {</span>
      const resultArr = <span class="cstat-no" title="statement not covered" >opts.expandKey ? ajax.response[opts.expandKey] : ajax.response</span>
<span class="cstat-no" title="statement not covered" >      return Array.isArray(resultArr) ? from(resultArr) : of(resultArr)</span>
    })
  )
}
&nbsp;
/** @description Turns a stream of objects into a stream of the patches between them.
 */
export const jsonPatch = <span class="fstat-no" title="function not covered" >() =&gt; <span class="cstat-no" title="statement not covered" ><span class="fstat-no" title="function not covered" ></span>&lt;T&gt;(</span>source: Observable&lt;T&gt;) =&gt;</span>
<span class="cstat-no" title="statement not covered" >  new Observable&lt;Operation[]&gt;(<span class="fstat-no" title="function not covered" >observer =&gt; {</span></span>
    let lastSeen: any = <span class="cstat-no" title="statement not covered" >{}</span>
&nbsp;
<span class="cstat-no" title="statement not covered" >    return source.subscribe({</span>
      next<span class="fstat-no" title="function not covered" >(</span>newObj) {
        const patch = <span class="cstat-no" title="statement not covered" >compare(lastSeen, newObj)</span>
<span class="cstat-no" title="statement not covered" >        if (patch.length === 0) <span class="cstat-no" title="statement not covered" >return</span></span>
&nbsp;
        // TODO allow for tests to be included in the patches that assert the old values
<span class="cstat-no" title="statement not covered" >        observer.next(patch)</span>
<span class="cstat-no" title="statement not covered" >        lastSeen = newObj</span>
      },
      error<span class="fstat-no" title="function not covered" >(</span>err) {
<span class="cstat-no" title="statement not covered" >        observer.error(err)</span>
      },
      complete<span class="fstat-no" title="function not covered" >() {</span>
<span class="cstat-no" title="statement not covered" >        observer.complete()</span>
      }
    })
  })
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Tue Sep 04 2018 14:08:26 GMT-0500 (CDT)
</div>
</div>
<script src="../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../sorter.js"></script>
<script src="../block-navigation.js"></script>
</body>
</html>
