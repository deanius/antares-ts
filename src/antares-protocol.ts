import { Subject, Observable, Subscription, Scheduler } from 'rxjs'

// Import here Polyfills if needed. Recommended core-js (npm i -D core-js)

export interface Action {
  type: string
  payload?: any
  error?: boolean
  meta?: Object
}

export interface ActionStreamItem {
  action: Action
}

export interface Renderer {
  (item: ActionStreamItem): any
}

export enum RenderMode {
  sync = 'sync',
  async = 'async'
}

export interface RendererConfig {
  mode: RenderMode
}

export default class AntaresProtocol {
  subject: Subject<ActionStreamItem>
  action$: Observable<ActionStreamItem>

  constructor() {
    this.subject = new Subject<ActionStreamItem>()
    this.action$ = this.subject.asObservable()
  }

  process(action: Action): Promise<ActionStreamItem> {
    this.subject.next({ action })
    return Promise.resolve({ action })
  }

  subscribeRenderer(
    renderer: Renderer,
    config: RendererConfig = { mode: RenderMode.sync }
  ): Subscription {
    const subscribeTo =
      config.mode === RenderMode.async ? this.action$.observeOn(Scheduler.async) : this.action$
    return subscribeTo.subscribe(renderer)
  }
}
