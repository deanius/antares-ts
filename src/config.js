// Configs that can be overridden by AntaresInit
export const Agents = {
  server: () =>
    typeof process !== "undefined" &&
    process.env &&
    process.env.NODE_ENV &&
    process.env.NODE_ENV !== "browser" &&
    !(typeof Meteor === "object" && Meteor.isClient),
  client: () => !Agents.server()
}
export const Epics = {}
export const NewId = [() => Math.floor(Math.random() * 10000000).toString(16)]
export const ParentAgentId = []
export const Types = {}
export const ReducerForKey = []
export const ViewReducer = []
export const MetaEnhancers = [
  // like airplane flight numbers, only guaranteed unique within small time window
  iAction =>
    iAction.getIn(["meta", "antares", "actionId"])
      ? {}
      : { actionId: Math.floor(Math.random() * 10000000).toString(16) },
  (_, Antares) => ({ originAgentId: Antares.agentId })
]
export const DispatchProxy = []
