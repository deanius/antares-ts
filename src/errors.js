export function ReductionError(err) {
  this.error = err
}
ReductionError.prototype = Object.create(Error.prototype)

export function KeyLookupFailed(err) {
  this.error = err
}
KeyLookupFailed.prototype = Object.create(Error.prototype)

export function ParentNotificationError(err) {
  this.error = err
}
ParentNotificationError.prototype = Object.create(Error.prototype)

export function RenderError(err) {
  this.error = err
}
RenderError.prototype = Object.create(Error.prototype)
RenderError.prototype.constructor = RenderError
