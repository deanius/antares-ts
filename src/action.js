import { fromJS, Map as iMap } from "immutable"
import Rx from "rxjs"

import { MetaEnhancers } from "./config"

let { Observable } = Rx

export const enhanceActionMeta = (action, oneTimeMetaEnhancer, Antares) => {
  let iAction = fromJS(action || {}).updateIn(["meta", "antares"], p => {
    return p || new iMap()
  })

  let enhancers
  if (typeof oneTimeMetaEnhancer === "function") {
    enhancers = MetaEnhancers.concat(oneTimeMetaEnhancer)
  } else {
    enhancers = MetaEnhancers
  }

  // apply meta enhancers, making sure they dont mess with anything
  let enhancedIAction = enhancers.reduce((iAction, enhancer) => {
    if (typeof enhancer !== "function") {
      debugger
    }
    let newMeta = enhancer(iAction, Antares)
    return iAction.mergeIn(["meta", "antares"], newMeta)
  }, iAction)

  return enhancedIAction.toJS()
}

export const createConsequence = (parent, consequence) => {
  let { type, payload, meta } = consequence
  let antaresMeta = (meta || {}).antares
  let parentMeta = (parent.meta && parent.meta.antares) || {}

  return {
    type,
    payload,
    meta: {
      ...meta,
      antares: {
        ...parentMeta,
        ...antaresMeta,
        actionId: Math.floor(Math.random() * 1000000).toString(12),
        parentActionId: parent.meta.antares.actionId
      }
    }
  }
}

export const localConsequence = (parent, consequence) => {
  let { type, payload, meta } = consequence
  return createConsequence(parent, {
    type,
    payload,
    meta: {
      ...meta,
      antares: {
        ...(meta || {}).antares,
        localOnly: true
      }
    }
  })
}

const begins = action =>
  createConsequence(action, {
    type: `${action.type}.begin`,
    payload: action.payload,
    meta: {
      antares: { localOnly: true }
    }
  })
const ends = (action, result) =>
  createConsequence(action, {
    type: `${action.type}.end`,
    payload: result,
    meta: {
      antares: {
        localOnly: true,
        concludesEpic: action.meta.antares.actionId
      }
    }
  })
const errors = (action, err) =>
  createConsequence(action, {
    type: `${action.type}.error`,
    payload: err,
    meta: {
      antares: { localOnly: true }
    }
  })
const concludes = (action, promise) =>
  Observable.fromPromise(
    promise
      .then(result => ends(action, result))
      .catch(err => errors(action, err))
  )
export const createPromiseEpic = (type, promiseFactory) => action$ =>
  action$
    .ofType(type)
    .mergeMap(action =>
      Observable.of(begins(action)).concat(
        concludes(action, promiseFactory(action))
      )
    )
