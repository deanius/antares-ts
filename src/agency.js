import { Agents } from "./config"

export const isInAgency = agencyType => {
  let runOnThisAgent = Agents[agencyType] || (() => false)
  return runOnThisAgent()
}

export const inAgencyRun = (givenType, fn) => {
  const agencyType = givenType === "*" ? "any" : givenType
  isInAgency(agencyType) && fn.call(this || global)
}
