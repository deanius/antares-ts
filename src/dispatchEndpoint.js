import { ReductionError } from "./errors"
import { logger } from "./logger"

// Defines the upstream (aka server) implementation of dispatch which:
//   dispatches to local store
//   acks the client
//   and that is all.
// The publication of the dispatched, and consequent actions is done by a final middleware
// that runs after the epic middleware to ensure that every listener potentially can hear
export const dispatchEndpoint = store => action => {
  logger.log(`New action ${action.type}`, {
    prefix: `AE (${action.meta.antares.actionId})`,
    newSection: true
  })

  // Now attempt to dispatch the action to the local store.
  // Any renderers that have been attached synchronously will run in the order subscribed.
  // Any exception in a synchronous renderer will blow the stack here and cause the store's
  // contents to remain unchanged. We'll need to resubscribe in that case.
  try {
    store.dispatch(action)
  } catch (ex) {
    logger.log(ex, {
      prefix: `AD Reduction Error (${action.meta.antares.actionId})`
    })
    throw new ReductionError(ex)
  }

  // TODO return results of all sync renderers
  // return action.meta.result
}
