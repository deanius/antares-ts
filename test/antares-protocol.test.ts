import { AntaresProtocol, Action, ActionStreamItem, RenderMode } from '../src/antares-protocol'
import fs from 'fs'
import { default as faker } from 'faker'
import { Observable, Subject, BehaviorSubject } from 'rxjs'
import { map, skip, take, toArray } from 'rxjs/operators'
import { debug } from 'util'

// wraps an it scenario and silences console messages during its executions
const inSilence = itFn => {
  const callIt = done => {
    const _console = global.console
    global.console = { log: jest.fn(), error: jest.fn() } // tslint:disable-line
    try {
      itFn(done)
    } catch (ex) {
      // tslint:disable-line
    } finally {
      global.console = _console
    }
  }

  // preserve arity in returned fn
  return itFn.length === 1 ? done => callIt(done) : () => callIt(undefined)
}
// Sanity check
describe('AntaresProtocol', () => {
  it('is instantiable', () => {
    expect(new AntaresProtocol()).toBeInstanceOf(AntaresProtocol)
  })
  it('has process, subscribeRenderer methods', () => {
    const antares = new AntaresProtocol()
    expect(antares).toMatchObject({
      process: expect.any(Function),
      subscribeRenderer: expect.any(Function)
    })
  })

  describe('instance methods', () => {
    let antares: AntaresProtocol

    beforeEach(() => {
      antares = new AntaresProtocol()
    })

    describe('#process', () => {
      it('accepts an action, and returns a promise', () => {
        const result = antares.process({ type: 'SomeAction' })
        expect(result).toBeInstanceOf(Promise)
      })
    })

    describe('#subscribeRenderer', () => {
      it('accepts a function to be called upon processing of actions', () => {
        const renderer = jest.fn()
        const result = antares.subscribeRenderer(renderer)
        const action = { type: 'Rando' }

        antares.process(action)
        expect(renderer).toHaveBeenCalledWith(expect.objectContaining({ action }))
      })

      it('accepts { name:String } on the config object', () => {
        antares.subscribeRenderer(noRender, { name: 'JohnnyNull' })
        expect(Array.from(antares._rendererSubs.keys())).toMatchSnapshot()
      })
      it("defaults naming a renderer 'renderer_N'", () => {
        antares.subscribeRenderer(noRender, { name: 'JohnnyNull' })
        antares.subscribeRenderer(noRender)
        antares.subscribeRenderer(noRender, { name: 'JohnnyFive' })
        const rendererNames = Array.from(antares._rendererSubs.keys())
        expect(rendererNames).toContain('renderer_2')
        expect(rendererNames).toMatchSnapshot()
      })

      describe('synchronous (online) mode', () => {
        describe('a renderer error', () => {
          it('propogates up to the caller of #process', () => {
            antares.subscribeRenderer(() => {
              throw new Error('unconditionally')
            })
            const doIt = () => {
              antares.process({ type: 'any' })
            }
            expect(doIt).toThrowErrorMatchingSnapshot()
          })
          it(
            [
              'prevents subsequent renderers from running during that event loop turn',
              'unsubscribes the offending renderer',
              'allows other renderers to run in future event loop turns'
            ].join(', '),
            () => {
              let before = jest.fn()
              let after = jest.fn()
              expect.assertions(5)

              antares.subscribeRenderer(before)
              antares.subscribeRenderer(() => {
                throw new Error('unconditionally')
              })
              antares.subscribeRenderer(after)

              // force error - in this event loop turn some renderers wont run
              try {
                antares.process({ type: 'any' })
              } catch (ex) {
                expect(before).toHaveBeenCalled()
                expect(after).not.toHaveBeenCalled()
              }

              const shouldntThrow = () => {
                antares.process({ type: 'any' })
              }

              // Process again, and the 2 non-throwing renderers are still subscribed
              expect(shouldntThrow).not.toThrow()
              expect(before).toHaveBeenCalledTimes(2)
              expect(after).toHaveBeenCalledTimes(1)

              // Long story short - synchronous render errors should have their own exception handling.
              // You should not have to wrap process in try{ } - rather, handle errors Promise-style.
            }
          )
        })
      })

      describe('async (batch) mode', () => {
        describe('a renderer error', () => {
          it('does not propogate to the caller of #process', () => {
            antares.subscribeRenderer(
              () => {
                throw new Error('iShouldJustLogNotGoNuclear')
              },
              { mode: RenderMode.async }
            )

            const doIt = () => {
              antares.process({ type: 'any' })
            }

            expect(doIt).not.toThrow()
          })
        })
      })
    })

    describe('#process', () => {
      it('returns a resolved promise', () => {
        expect.assertions(1)
        const result = antares.process({ type: 'rando' })
        return expect(result).resolves.toBeTruthy()
      })

      it('can be used to abstract the WHAT from the HOW', () => {
        const bizTalk = faker.fake('{{company.bsBuzz}} {{company.bsAdjective}} {{company.bsNoun}}')

        expect.assertions(1)

        const action = {
          type: 'File.append',
          payload: {
            fileName: 'antares.test.log',
            content: bizTalk
          }
        }

        antares.subscribeRenderer(logFileAppender)
        antares.process(action)

        // Because we subscribed our renderer synchronously, we can expect the
        // text was written to the file.
        expect(fs.readFileSync(action.payload.fileName, 'UTF8')).toMatch(bizTalk)
      })
    })

    describe('#action$ - the action stream', () => {
      it('exposes each processed action', () => {
        expect.assertions(1)
        const randomActions = [{ type: 'rando 1' }, { type: 'rando 2' }]

        // get a promise for all seen actions from now, as an array
        const lastTwoActions = antares.action$
          .pipe(take(2), map(justTheAction), toArray())
          .toPromise()

        // process actions
        randomActions.forEach(a => antares.process(a))

        // expect the resolved value toEqual our randomActions
        return expect(lastTwoActions).resolves.toMatchSnapshot()
      })

      it('contains an entry on {results} for each synchronous renderer', () => {
        expect.assertions(4)

        // set up a listener for testing - a Subject whose .value is the last thing it observed
        const lastItemSubject = new BehaviorSubject<ActionStreamItem>(null)
        antares.action$.subscribe(lastItemSubject)

        // set up a normal renderer
        antares.subscribeRenderer(item => noSpecialValue, {
          name: 'rememberMyName',
          mode: RenderMode.sync
        })

        // I have to admit I dont understand why '2' below - something to do w/ BehaviorSubject
        const nextItem = lastItemSubject.pipe(take(2)).toPromise()
        // process an item
        antares.process({ type: 'fooz' })

        return nextItem.then(item => {
          expect(item).toHaveProperty('results')
          expect(item.results).toBeInstanceOf(Map)
          expect(Array.from(item.results.keys())).toContain('rememberMyName')
          expect(item).toMatchSnapshot()
        })
      })

      it('contains an entry on {resultsAsync} for each async renderer', () => {
        expect.assertions(4)

        // set up a listener for testing - a Subject whose .value is the last thing it observed
        const lastItemSubject = new BehaviorSubject<ActionStreamItem>(null)
        antares.action$.subscribe(lastItemSubject)

        // set up a normal renderer
        antares.subscribeRenderer(item => noSpecialValue, {
          name: 'asyncYo',
          mode: RenderMode.sync
        })

        // I have to admit I dont understand why '2' below - something to do w/ BehaviorSubject
        const nextItem = lastItemSubject.pipe(take(2)).toPromise()
        // process an item
        antares.process({ type: 'fooz' })

        return nextItem.then(item => {
          expect(item).toHaveProperty('resultsAsync')
          expect(item.results).toBeInstanceOf(Map)
          expect(Array.from(item.results.keys())).toContain('asyncYo')
          expect(item).toMatchSnapshot()
        })
      })
    })
  })
})

/**** util functions below ****/
const justTheAction = ({ action }: ActionStreamItem) => action
const noRender = () => null
const noSpecialValue = 'noSpecialValue'

const logFileAppender = ({ action: { type, payload } }) => {
  // Most renderers care about a subset of actions. Return early if you don't care.
  if (!type.match(/^File\./)) return

  const { fileName, content } = payload
  return fs.appendFileSync(fileName, content + '\n', { encoding: 'UTF8' })
}
