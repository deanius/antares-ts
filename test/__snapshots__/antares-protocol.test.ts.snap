// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AntaresProtocol instance methods #action$ - the action stream contains an entry on {results} for each synchronous renderer 1`] = `
Object {
  "action": Object {
    "type": "any",
  },
  "results": Map {
    "rememberMyName" => "syncReturnValue",
  },
  "resultsAsync": Map {},
}
`;

exports[`AntaresProtocol instance methods #action$ - the action stream exposes each processed action 1`] = `
Array [
  Object {
    "type": "rando 1",
  },
  Object {
    "type": "rando 2",
  },
]
`;

exports[`AntaresProtocol instance methods #subscribeRenderer accepts { name:String } on the config object 1`] = `
Array [
  "JohnnyNull",
]
`;

exports[`AntaresProtocol instance methods #subscribeRenderer async (batch) mode happy path will send returned actions back through #process 1`] = `
Array [
  Object {
    "type": "any",
  },
  Object {
    "type": "consequentialAction",
  },
]
`;

exports[`AntaresProtocol instance methods #subscribeRenderer defaults naming a renderer 'renderer_N' 1`] = `
Array [
  "JohnnyNull",
  "renderer_2",
  "JohnnyFive",
]
`;

exports[`AntaresProtocol instance methods #subscribeRenderer synchronous (online) mode a renderer error propogates up to the caller of #process 1`] = `"unconditionally"`;
